### ISSUE TITLE
RESTCommand responses not flexible enough

### ISSUE THREAD

**Issue by bclindner** [bclindner/valerius#6]
I wrote a primitive Pokedex command based on the RESTCommand. It gets the ID, name, types, and main sprite of the Pokemon as stored by the PokeAPI:
```json
    {
      "name": "Pokedex by name/number",
      "type": "rest",
      "options": {
        "triggerregex": "^!pokemon ([\\-a-z]+|[0-9]+)$",
        "endpoint": [
          "https://pokeapi.co/api/v2/pokemon/%s/",
          1
        ],
        "response": [
          "#%v %s (%s/%s): %s",
          "id",
          "name",
          "types.0.type.name",
          "types.1.type.name",
          "sprites.front_default"
        ]
      }
    },
```

This works fine for the most part - sending `!pokemon bulbasaur` returns what it should:
```
#1 bulbasaur (poison/grass): https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png
```

But in situations where a single-type pokemon is sent, like `!pokemon poochyena`, it'll throw a formatting type error:
```
#261 poochyena (dark/%!s(<nil>)): https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/261.png
```
Trying to fix this goes *way* beyond the scope of the `fmt` package. I think the most elegant solution here would be some kind of polymorphic response object that can adapt to that, but I haven't the first idea how to design such a thing right now.

**Comment by bclindner** [comment]
Maybe I'm overthinking this. Proposal: an array of response objects, where the command runner uses the first one that works?
Something like:
```json
        "responses": [
          [
            "#%v %s (%s/%s): %s",
            "id",
            "name",
            "types.0.type.name",
            "types.1.type.name",
            "sprites.front_default"
          ],
          [
            "#%v %s (%s): %s",
            "id",
            "name",
            "types.0.type.name",
            "sprites.front_default"
          ]
        ]
```
The runner would check if `types.1.type.name` exists, sees it doesn't, and then tries the other response object. If it has no types, it would then just ignore the command.

**Comment by bclindner** [comment]
Yeah, we're going with that for now. May change in the next major release.

**Comment by bclindner** [comment]
Done (06dd1d2).

**Comment by bclindner** [comment]
New idea - the `Printf` formatting is good for simple stuff, but `text/template` might give even more flexibility! Passing the JSON response into a user-defined template would mean you could loop over results, use flow control, and plenty more, on top of striking another 3rd party dependency off the list!

This would also address #5 which still kinda bugs me. I'm probably going to go ahead and try this at some point soon.

**Comment by bclindner** [comment]
I realize now that `text/template` is somewhat longwinded compared to the `Printf` solution, and this doesn't lend itself well to JSON (since JSON doesn't allow literal newlines in quoted strings). If this happens, it would be best to have a `responseTemplateFile` field or something similar that just holds a path to a file to parse, so that templates can be better managed.

**Comment by bclindner** [comment]
Rough implementation at aefdab1. This is the example configuration tested, a simple weather command:
```json
    {
      "name": "OpenWeatherMap",
      "type": "rest",
      "options": {
        "triggerregex": "^t!weather ([0-9]{5})$",
        "endpoint": [
          "https://api.openweathermap.org/data/2.5/weather?units=metric&appid=...&zip=%s",
          1
        ],
        "method": "GET",
        "response": "Weather in {{.name}}: {{(index .weather 0).main}}, {{.main.temp}}°C. High {{.main.temp_max}}°C, low {{.main.temp_min}}°C.",
        "errorMessage": "Couldn't find that one. Sorry!"
      }
    },

```
For contrast, the equivalent `Printf` version:
```json
    {
      "name": "OpenWeatherMap",
      "type": "rest",
      "options": {
        "triggerregex": "^v!weather ([0-9]{5})$",
        "endpoint": [
          "https://api.openweathermap.org/data/2.5/weather?units=metric&appid=6524e5e5578e5f7d75b431c302bf803b&zip=%s",
          1
        ],
        "method": "GET",
        "response": [
          "Weather in %v: %v, %v°C. %v°C high/%v°C low.",
          "name",
          "weather.0.main",
          "main.temp",
          "main.temp_min",
          "main.temp_max"
        ],
        "errorMessage": "Couldn't find that one. Sorry!"
      }
    },
```
Longwinded is right, but this is a much cleaner implementation that makes a little more sense. That said, the `{{(index .weather 0).main}}` is... *not* clean. Uncertain if I should be using `text/template` or perhaps another templating engine. Regardless, still implementing as proposed.

**Comment by bclindner** [comment]
Implemented response files (914bdab). Finicky, but *extremely* flexible compared to the `Printf` solution. I think I'll be merging this into master - I'd rather wrestle with template files than `Printf`commands and response fallthroughs any day.
